### PR TITLE
WeightInfo for Vesting Pallet

### PR DESCRIPTION
**Pull request by shawntabrizi**
polkadot companion: https://github.com/paritytech/polkadot/pull/1721

This PR introduces `WeightInfo` for the Vesting Pallet using the automated benchmarking system.

I added a `MaxLocks` configuration trait which allows the runtime engineer to specify the assumed max number of locks for any user. This value is not enforced by the pallet, but used for estimating weights. When running benchmarks, this value is used for the range samples taken for benchmarks.

All DB Reads/Writes are consistent, but there is a notable increase in the average weight of each function. But the value is consistent and what we have been seeing when upgrading other weights.

### CODE REVIEW

File: bin/node/runtime/src/lib.rs [hunk 1]
```diff
@@ -849,14 +849,18 @@ impl pallet_society::Trait for Runtime {
 
 parameter_types! {
 	pub const MinVestedTransfer: Balance = 100 * DOLLARS;
+	// For weight estimation, we assume that the most locks on an individual account will be 50.
+	// This number may need to be adjusted in the future if this assumption no longer holds true.
+	pub const MaxLocks: u32 = 50;
```
Comment (kianenigma) [inline bin/node/runtime/src/lib.rs:854]: I don't think this makes much sense. Reading this, I assume that if I set this, then the maximum locks are limited to 50, but this is not the case. Setting this to 5 does not limit anything. 

If we actually need to have a limit on the number of locks, then if there is a way for it, it makes a lot more sense to put this config as `type` in system trait or as `const` in `Currency` trait. 

Another reason is that I am sure there will be (and already is) more modules that have operations on lock, and therefore their benchmarks will need some form of `MaxLocks`.

If there is no need for the system to make this assumption, then I think it is fine to just make a constant that is large enough and do the same `let l in 0 .. MAX_LOCKS;` If `MAX_LOCK` is large enough, then the formula won't change regardless of what `MAX_LOCK` is. 

---

File: bin/node/runtime/src/weights/pallet_vesting.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+//! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 2.0.0-rc6
+
+#![allow(unused_parens)]
+#![allow(unused_imports)]
+
+use frame_support::weights::{Weight, constants::RocksDbWeight as DbWeight};
+
+pub struct WeightInfo;
+impl pallet_vesting::WeightInfo for WeightInfo {
```
Comment (kianenigma) [inline bin/node/runtime/src/weights/pallet_vesting.rs:26]: as nudged to myself as well, need to be implemented for `()`, or did we re-consider? :p

Comment (kianenigma) [reply bin/node/runtime/src/weights/pallet_vesting.rs:26]: ahh this file is probably need to be removed.

File: frame/vesting/src/lib.rs [hunk 3]
```diff
@@ -238,13 +234,10 @@ decl_module! {
...
-		/// - Benchmark:
-		///     - Unlocked: 44.3 + .294 * l µs (min square analysis)
-		///     - Locked: 48.16 + .103 * l µs (min square analysis)
-		/// - Using 50 µs fixed. Assuming less than 50 locks on any user, else we may want factor in number of locks.
 		/// # </weight>
-		#[weight = 50_000_000 + T::DbWeight::get().reads_writes(3, 3)]
-		fn vest_other(origin, target: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+		#[weight = T::WeightInfo::vest_other_locked(T::MaxLocks::get())
+			.max(T::WeightInfo::vest_other_unlocked(T::MaxLocks::get()))
+		]		fn vest_other(origin, target: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
```
Comment (kianenigma) [inline frame/vesting/src/lib.rs:240]:
```suggestion
		]		
		fn vest_other(origin, target: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
```

File: frame/vesting/src/lib.rs [hunk 4]
```diff
@@ -473,12 +462,14 @@ mod tests {
...
 	parameter_types! {
 		pub const MinVestedTransfer: u64 = 256 * 2;
+		pub const MaxLocks: u32 = 10;
 	}
 	impl Trait for Test {
 		type Event = ();
 		type Currency = Balances;
 		type BlockNumberToBalance = Identity;
 		type MinVestedTransfer = MinVestedTransfer;
+		type MaxLocks = MaxLocks;
```
Comment (kianenigma) [inline frame/vesting/src/lib.rs:472]:
```suggestion
		type MaxLocks = ();
```

I think implementation of `Get` for `()` is underrated in our repo :D.

File: bin/node/runtime/src/lib.rs (hunk 1 above)
Comment (shawntabrizi) [reply bin/node/runtime/src/lib.rs:854]: Having some hard limit in the balances module is hard because it would ultimately change the API around locks, where adding a lock may return a result, and thus any logic which adds a lock may need to now handle this error, and it could be quite an impact across the runtime.

As for "selecting a large number", that is basically what I have done here, except as I have learned more consistently over time, we should make this configurable. A "big number" for us may not be a big number for someone else. Or maybe our choice is crazy large for some other runtime.

Basically, all the functions in vesting are some `f(max_locks)`, and so we let them define those functions, and we also let them define the `max_locks` variable.

File: bin/node/runtime/src/weights/pallet_vesting.rs (hunk 2 above)
Comment (shawntabrizi) [reply bin/node/runtime/src/weights/pallet_vesting.rs:26]: What do you mean here?

I implement `()` in the `default_weight.rs` file, but I also define it here because it illustrates how weights should be used for individual runtimes. A runtime should not depend on `()` unless it is a test runtime since it can change from upstream.

Comment (kianenigma) [reply bin/node/runtime/src/weights/pallet_vesting.rs:26]: yes, my misunderstanding.

File: bin/node/runtime/src/lib.rs (hunk 1 above)
Comment (thiolliere) [reply bin/node/runtime/src/lib.rs:854]: maybe we can change the name to estimated_max_lock ? or weight_for_locks ?

but I agree that if vesting needs MaxLocks then some other pallet will need it as well.

Comment (shawntabrizi) [reply bin/node/runtime/src/lib.rs]: Updated per discussion. A new trait is introduced into the Balances Pallet and the `LockableCurrency` trait. And this is ingested in the pallet.

File: frame/balances/src/lib.rs [hunk 5]
```diff
@@ -663,6 +672,12 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
 
 	/// Update the account entry for `who`, given the locks.
 	fn update_locks(who: &T::AccountId, locks: &[BalanceLock<T::Balance>]) {
+		if locks.len() as u32 > T::MaxLocks::get() {
+			frame_support::print(
```
Comment (kianenigma) [inline frame/balances/src/lib.rs:676]: This will go into log target `runtime` and `debug` level I think? Better to make it warn. I think you can use `frame_support::debug::warn!(...)`.

File: frame/vesting/src/lib.rs [hunk 6]
```diff
@@ -463,7 +450,11 @@ mod tests {
 		type OnKilledAccount = ();
 		type SystemWeightInfo = ();
 	}
+	parameter_types! {
+		pub const MaxLocks: u32 = 10;
+	}
```
Comment (kianenigma) [inline frame/vesting/src/lib.rs:455]:
```suggestion
```

**Review by kianenigma — APPROVED** [review]
Looks good module using `warn!`.

File: frame/balances/src/lib.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/balances/src/lib.rs:676]:
```suggestion
			frame_support::debug::warn!(
```

**Comment by shawntabrizi** [comment]
bot merge
